src/nbl/ui/CWindowWin32.cpp

```diff
@@ -8,276 +8,70 @@ namespace ui
-			if (event.KeyInput.Key == nbl::KEY_SHIFT)
-			{
-				// this will fail on systems before windows NT/2000/XP, not sure _what_ will return there instead.
-				event.KeyInput.Key = (nbl::EKEY_CODE)MapVirtualKey(((lParam >> 16) & 255), MY_MAPVK_VSC_TO_VK_EX);
-			}
-			if (event.KeyInput.Key == nbl::KEY_CONTROL)
-			{
-				event.KeyInput.Key = (nbl::EKEY_CODE)MapVirtualKey(((lParam >> 16) & 255), MY_MAPVK_VSC_TO_VK_EX);
-				// some keyboards will just return LEFT for both - left and right keys. So also check extend bit.
-				if (lParam & 0x1000000)
-					event.KeyInput.Key = nbl::KEY_RCONTROL;
-			}
-			if (event.KeyInput.Key == nbl::KEY_MENU)
-			{
-				event.KeyInput.Key = (nbl::EKEY_CODE)MapVirtualKey(((lParam >> 16) & 255), MY_MAPVK_VSC_TO_VK_EX);
-				if (lParam & 0x1000000)
-					event.KeyInput.Key = nbl::KEY_RMENU;
-			}
-
-			GetKeyboardState(allKeys);
-
-			event.KeyInput.Shift = ((allKeys[VK_SHIFT] & 0x80) != 0);
-			event.KeyInput.Control = ((allKeys[VK_CONTROL] & 0x80) != 0);
-
-			// Handle unicode and deadkeys in a way that works since Windows 95 and nt4.0
-			// Using ToUnicode instead would be shorter, but would to my knowledge not run on 95 and 98.
-			WORD keyChars[2];
-			UINT scanCode = HIWORD(lParam);
-			int conversionResult = ToAsciiEx(wParam, scanCode, allKeys, keyChars, 0, KEYBOARD_INPUT_HKL);
-			if (conversionResult == 1)
+			uint32_t newWidth = LOWORD(lParam);
+			uint32_t newHeight = HIWORD(lParam);
+			eventCallback->onWindowResized(window, newWidth, newHeight);
+			switch (wParam)
 			{
-				WORD unicodeChar;
-				MultiByteToWideChar(
-					KEYBOARD_INPUT_CODEPAGE,
-					MB_PRECOMPOSED, // default
-					(LPCSTR)keyChars,
-					sizeof(keyChars),
-					(WCHAR*)&unicodeChar,
-					1);
-				event.KeyInput.Char = unicodeChar;
+			case SIZE_MAXIMIZED:
+				eventCallback->onWindowMaximized(window);
+				break;
+			case SIZE_MINIMIZED:
+				eventCallback->onWindowMinimized(window);
+				break;
 			}
-			else
-				event.KeyInput.Char = 0;
-
-			// allow composing characters like '@' with Alt Gr on non-US keyboards
-			if ((allKeys[VK_MENU] & 0x80) != 0)
-				event.KeyInput.Control = 0;
-
-			dev = getDeviceFromHWnd(hWnd);
-			if (dev)
-				dev->postEventFromUser(event);
-
-			if (message == WM_SYSKEYDOWN || message == WM_SYSKEYUP)
-				return DefWindowProc(hWnd, message, wParam, lParam);
-			else
-				return 0;
+			break;
 		}
-
-		case WM_SIZE:
+		case WM_SETFOCUS:
 		{
-			// resize
-			dev = getDeviceFromHWnd(hWnd);
-			if (dev)
-				dev->OnResized();
+			eventCallback->onGainedKeyboardFocus(window);
+			break;
 		}
-		return 0;
-
-		case WM_DESTROY:
-			PostQuitMessage(0);
-			return 0;
-
-		case WM_SYSCOMMAND:
-			// prevent screensaver or monitor powersave mode from starting
-			if ((wParam & 0xFFF0) == SC_SCREENSAVE ||
-				(wParam & 0xFFF0) == SC_MONITORPOWER ||
-				(wParam & 0xFFF0) == SC_KEYMENU
-				)
-				return 0;
-
+		case WM_KILLFOCUS:
+		{
+			eventCallback->onLostKeyboardFocus(window);
```

Comment:
why is it called keyboard focus?

Reply:
hmmm. I'd probably make a single callback `onFocusChange(bool gained, E_FOCUS_TYPE_FLAGS flags)` for all this